### PR TITLE
[doc][trivial] Add basic git squash workflow [skip ci]

### LINKED ISSUE #8032: Add 'how to squash several fixup commits' to our coding guideline
Adding this to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md helps to not have it copy paste in every comment: https://github.com/bitcoin/bitcoin/pull/8004#issuecomment-217810102

### PR DESCRIPTION
**Pull request by fanquake**
Fixes #8032

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -35,13 +35,21 @@ The title of the pull request should be prefixed by the component or area that t
...
 
-At this stage one should expect comments and review from other contributors. You can add more commits to your pull request by committing them locally and pushing to your fork until you have satisfied all feedback. If your pull request is accepted for merging, you may be asked by a maintainer to squash and or rebase your commits before it will be merged. The length of time required for peer review is unpredictable and will vary from patch to patch.
+At this stage one should expect comments and review from other contributors. You can add more commits to your pull request by committing them locally and pushing to your fork until you have satisfied all feedback. If your pull request is accepted for merging, you may be asked by a maintainer to squash and or [rebase](https://git-scm.com/docs/git-rebase) your commits before it will be merged. The basic squashing workflow is shown below.
+
+    git checkout your_branch_name
+    git rebase -i commit_shasum
```
Comment (MarcoFalke) [inline CONTRIBUTING.md]: What is `commit_shasum`?

Maybe `HEAD~n`?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -35,13 +35,21 @@ The title of the pull request should be prefixed by the component or area that t
...
+    # set commits from 'pick' to 'squash'
+    # save and quit the editor
+    git push -f (force push to GitHub)
```
Comment (MarcoFalke) [inline CONTRIBUTING.md]: nit: missing `#` after `-f`. (or leave the comment away)

**Comment by MarcoFalke** [comment]
It might be worth to mention that squashing is not required when the commits each have a different "topic"/task and are not fixup commits.

Also, it might come in handy to add section headings, so it is easier to link to a section...


Concept ACK.

File: CONTRIBUTING.md (hunk 1 above)
Comment (jonasschnelli) [inline CONTRIBUTING.md]: Maybe `git rebase -i HEAD~<amount of commits to squash>`

Maybe add something like (please clean/rewrite the english)
```
Use `p` at the beginning of the line to indicate the commit you wish to use as the base commit
Use `s` at the beginning of the line for all commits you want to squash into the commit you have marked with `p`.
```

**Comment by laanwj** [comment]
Concept ACK, this was long overdue, somehow I was not lazy enough describe this again in every new contributor's pull, but too lazy to add it into here :) Thanks for doing so.

File: CONTRIBUTING.md (hunk 1 above)
Comment (laanwj) [inline CONTRIBUTING.md]: > git rebase -i HEAD~<amount of commits to squash>

It's not the number of commits to squash, but at least that amount +1, as you need a base commit too to squash into.
Usually more like the total number of commits in someone's PR that they want to edit.

I haven't found a really user friendly way to determine this base commit yet, which is why I usually look it up for people. HEAD~x may be best, but is also something to get wrong easily if you miscount.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -35,13 +35,21 @@ The title of the pull request should be prefixed by the component or area that t
...
 
-At this stage one should expect comments and review from other contributors. You can add more commits to your pull request by committing them locally and pushing to your fork until you have satisfied all feedback. If your pull request is accepted for merging, you may be asked by a maintainer to squash and or rebase your commits before it will be merged. The length of time required for peer review is unpredictable and will vary from patch to patch.
+At this stage one should expect comments and review from other contributors. You can add more commits to your pull request by committing them locally and pushing to your fork until you have satisfied all feedback. If your pull request is accepted for merging, you may be asked by a maintainer to squash and or [rebase](https://git-scm.com/docs/git-rebase) your commits before it will be merged. The basic squashing workflow is shown below.
+
+    git checkout your_branch_name
+    git rebase -i commit_shasum
+    # set commits from 'pick' to 'squash'
+    # save and quit the editor
```
Comment (laanwj) [inline CONTRIBUTING.md]: After this, they'll get another editor in which they can combine/refine the commit messages.

File: CONTRIBUTING.md (hunk 1 above)
Comment (MarcoFalke) [inline CONTRIBUTING.md]: For me, git will tell me what the commands are when I am doing the rebase:

```
# Commands: 
# p, pick = use commit 
# r, reword = use commit, but edit the commit message 
# e, edit = use commit, but stop for amending 
# s, squash = use commit, but meld into previous commit 
# f, fixup = like "squash", but discard this commit's log message 
# x, exec = run command (the rest of the line) using shell

**Comment by laanwj** [comment]
BTW: the [skip ci] did'n't work. I think you need to put it in the commit message, unfortunately.
(doesn't have to be in the first line though, luckily)

**Comment by MarcoFalke** [comment]
utACK 90963e5

**Comment by paveljanik** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/8034/commits/90963e53707cf244e6a2a804341e9280628b0c34

**Comment by achow101** [comment]
utACK 90963e5

**Comment by petertodd** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/8034/commits/90963e53707cf244e6a2a804341e9280628b0c34

**Comment by paveljanik** [comment]
This is ready for merge.

**Comment by arowser** [comment]
Can one of the admins verify this patch?

**Pull request merged**